### PR TITLE
Stateful object

### PR DESCRIPTION
**Pull request by stunnerboyone**

### CODE REVIEW

File: src/transformState.js [hunk 1]
```diff
@@ -4,8 +4,32 @@
  * @param {Object} state
  * @param {Object[]} actions
  */
+
 function transformState(state, actions) {
   // write code here
+  for (let i = 0; i < actions.length; i++) {
```
Comment (denys-danyliuk) [inline src/transformState.js]: use for of

File: src/transformState.js [hunk 2]
```diff
@@ -4,8 +4,32 @@
...
+    switch (actions[i].type) {
+      case 'addProperties':
+        Object.assign(state, { ...actions[i].extraData });
+        break;
+      case 'removeProperties':
+        for (let j = 0; j < actions[i].keysToRemove.length; j++) {
```
Comment (denys-danyliuk) [inline src/transformState.js]: same here, it's unnecessary complicated

File: src/transformState.js [hunk 3]
```diff
@@ -4,8 +4,32 @@
...
+        break;
+      case 'clear':
+        for (const key in state) {
+          delete state[key];
+        }
+        break;
+      default:
+        const errorMessage = new Error('Array have not correct action value');
+
+        return errorMessage;
```
Comment (denys-danyliuk) [inline src/transformState.js]:
```suggestion
        throw new Error('Array have not correct action value');
```

File: src/transformState.js [hunk 4]
```diff
@@ -4,8 +4,32 @@
...
+    }
+  }
+
+  return state;
```
Comment (denys-danyliuk) [inline src/transformState.js]: function doesn't return anything, it just transforms state

**Review by denys-danyliuk — CHANGES_REQUESTED** [review]

File: src/transformState.js [hunk 5]
```diff
@@ -4,8 +4,28 @@
  * @param {Object} state
  * @param {Object[]} actions
  */
+
 function transformState(state, actions) {
   // write code here
+  for (const action in actions) {
+    switch (actions[action].type) {
```
Comment (andriy-shymkiv) [inline src/transformState.js]: `actions` is array. so which method we should use to run thorough the array?

```suggestion
  for (const action of actions) {
    switch (action.type) {
```

Comment (andriy-shymkiv) [reply src/transformState.js]: use this approach everywhere

File: src/transformState.js [hunk 6]
```diff
@@ -4,8 +4,28 @@
...
+      case 'addProperties':
+        Object.assign(state, { ...actions[action].extraData });
+        break;
+      case 'removeProperties':
+        for (const keyToRemove in actions[action].keysToRemove) {
+          delete state[actions[action].keysToRemove[keyToRemove]];
+        }
```
Comment (andriy-shymkiv) [inline src/transformState.js]: this loop is really complicated. after using a `for of` loop above -  you can make this one mutch simpler

**Review by andriy-shymkiv — CHANGES_REQUESTED** [review]
good job! left comments

File: src/transformState.js [hunk 7]
```diff
@@ -4,8 +4,28 @@
...
+
 function transformState(state, actions) {
   // write code here
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(state, { ...action.extraData });
+        break;
+      case 'removeProperties':
+        for (const keyToRemove of action.keysToRemove) {
```
Comment (andriy-shymkiv) [inline src/transformState.js:16]: looks better now, but `action` is object here, so why did you changed it to `for of`?
remember - use `for of` for arrays, and `for in` for objects

**Review by andriy-shymkiv — CHANGES_REQUESTED** [review]

File: src/transformState.js (hunk 7 above)
Comment (stunnerboyone) [reply src/transformState.js:16]: image
action is object but action.keysToRemove an array

**Review by xiimiik — APPROVED** [review]
GJ

**Pull request closed without merging**